Whitespace incorrectly trimmed from variables?

It feels like gosh trims leading and trailing spaces from variables:
```
$ cat x.sh
#!/bin/sh
FOO=" Q "
echo a${FOO}b
$ sh x.sh
a Q b
$ go run main.go x.sh
aQb
```

I would have expected gosh x.sh and sh x.sh to produce identical output.

The problem can be reproduced with this unit test case in interp/interp_test.go:
```
--- a/interp/interp_test.go
+++ b/interp/interp_test.go
@@ -310,6 +310,7 @@ var runTests = []runTest{
        {"echo  foo_interp_missing ", "foo_interp_missing\n"},
        {"echo ' foo_interp_missing '", " foo_interp_missing \n"},
        {`echo " foo_interp_missing "`, " foo_interp_missing \n"},
+       {`a=" foo_interp_missing "; echo $a`, " foo_interp_missing \n"},
        {`echo a'b'c"d"e`, "abcde\n"},
        {`a=" b c "; echo $a`, "b c\n"},
        {`a=" b c "; echo "$a"`, " b c \n"},
C02DG0QJMD6R:interp dkegel$ go test
--- FAIL: TestRunnerRun (0.04s)
    --- FAIL: TestRunnerRun/#67 (0.00s)
        interp_test.go:3328: wrong output in "a=\" foo_interp_missing \"; echo $a":
            want: " foo_interp_missing \n"
            got:  "foo_interp_missing\n"
```

--- a/expand/expand.go
+++ b/expand/expand.go
@@ -546,11 +546,22 @@ func (cfg *Config) wordFields(wps []syntax.WordPart) ([][]fieldPart, error) {
 		curField = nil
 	}
 	splitAdd := func(val string) {
-		for i, field := range strings.FieldsFunc(val, cfg.ifsRune) {
-			if i > 0 {
+		fieldStart := -1
+		for i, r := range val {
+			if cfg.ifsRune(r) {
+				if fieldStart >= 0 { // ending a field
+					curField = append(curField, fieldPart{val: val[fieldStart:i]})
+					fieldStart = -1
+				}
 				flush()
+			} else {
+				if fieldStart < 0 { // starting a new field
+					fieldStart = i
+				}
 			}
-			curField = append(curField, fieldPart{val: field})
+		}
+		if fieldStart >= 0 { // ending a field without IFS
+			curField = append(curField, fieldPart{val: val[fieldStart:]})
 		}
 	}
 	for i, wp := range wps {

--- a/interp/interp_test.go
+++ b/interp/interp_test.go
@@ -313,6 +313,8 @@ var runTests = []runTest{
 	{`echo a'b'c"d"e`, "abcde\n"},
 	{`a=" b c "; echo $a`, "b c\n"},
 	{`a=" b c "; echo "$a"`, " b c \n"},
+	{`a=" b c "; echo foo${a}bar`, "foo b c bar\n"},
+	{`a="b    c"; echo foo${a}bar`, "foob cbar\n"},
 	{`echo "$(echo ' b c ')"`, " b c \n"},
 	{"echo ''", "\n"},
 	{`$(echo)`, ""},

